_drafts/faster-builds.md

```diff
@@ -0,0 +1,649 @@
+the `ccache` binary and instead to set up your build system in the following
+way:
+
+```make
+ARM_CC    ?= arm-none-eabi-gcc
+CC_PREFIX ?= ccache
+CC        ?= $(CC_PREFIX) $(ARM_CC)
+
+%.o : %.c
+	$(CC) $(CFLAGS) -c -o $@ $<
+```
+
+With this setup, your environment remains simple and developers only need to
+install ccache with no further setup.
+
+You can easily check whether ccache is set up properly using the following steps
+and verifying there are "cache hits".
+
+```
+# This initial build with populate the cache
+$ make
+...
+
+# Clear ccache stats
+$ ccache --zero-stats
+
+# Run the build again (most artifacts should be cached)
+$ make clean && make
+...
+
+# Ensure ccache had some hits
+$ ccache --show-stats
+cache directory                     /Users/tyler/.ccache
+primary config                      /Users/tyler/.ccache/ccache.conf
+secondary config      (readonly)    /usr/local/Cellar/ccache/3.7.3/etc/ccache.conf
+stats updated                       Mon Feb 10 23:24:26 2020
+stats zeroed                        Mon Feb 10 23:24:18 2020
+cache hit (direct)                     0
+cache hit (preprocessed)              75
+cache miss                             0
+cache hit rate                    100.00 %
+called for link                        4
+can't use precompiled header           1
+cleanups performed                     0
+files in cache                      8521
+cache size                         197.0 MB
+max cache size                       5.0 GB
+```
+
+As you can see above, I had a 100% cache hit rate on a rebuild!
+
+> If you have a _very_ large project, I've heard good things about icecc[^2] and
+> sccache[^3], but these are generally for projects with thousands of large
+> files and many developers.
+
+### Optimizing Header Includes and Dependencies
+
+This (large) section goes into how to clean up and optimize dependencies between
+your source files and headers.
+
+#### Large Files = Problem
+
+Generally speaking, the larger the input file into the compiler is, the longer
+it will take to compile. If the compiler is fed a file with 10 lines, it might
+take 10 ms to compile, but with 5000 lines, it might take 100 ms.
+
+One step that happens before a file is compiled is "preprocessing". A
+preprocessor takes all `#include "file.h"` statements and replaces them with the
+_entire contents_ of the file being included, and does this recursively. This
+means that a `.c` file as small as the following could wind up being 5000+ lines
+of code for the compiler!
+
+```c
+// extra.c
+
+#include "stm32f4xx.h"
+
+char *extra(void) {
+  return "extra";
+}
+```
+
+We can check the contents of the preprocessed (post-processed) file by calling
+`gcc` with the `-E` flag.
+
+```
+$ arm-none-eabi-gcc -E -g -Os ... -c src/extra.c > preprocessed.txt
+$ cat preprocessed.txt | wc -l
+    5397
+```
+
+It turns out that the header file `stm32f4xx.h` and all of _its_ included header
```

Comment:
```suggestion
It turns out that the header file `stm32f4xx.h` and its transitively included header
```

Prefer words over visual styling for better reading flow

Reply:
Love that word choice. I couldn't figure out a good one so opted for emphasis.